account_invoice_update_stock_valuation/models/account_move.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2023 Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo import _, api, fields, models
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    svl_warn_msg = fields.Text(compute="_compute_svl_warn_msg")
+
+    @api.depends("invoice_line_ids.stock_valuation_ok")
+    def _compute_svl_warn_msg(self):
+        for invoice in self:
+            svl_warn_msg = False
+            if invoice.state == "draft" and invoice.invoice_line_ids.filtered(
+                lambda l: not l.stock_valuation_ok
+            ):
+                svl_warn_msg = _(
+                    "Warning: in this invoice, there are lines that have a unit price "
+                    "that is different from that of the purchase order ! "
+                    "The stock valuation as well as the purchase price of the product "
+                    "will be updated when the invoice is validated."
+                )
+            invoice.svl_warn_msg = svl_warn_msg
+
+    def action_post(self):
+        res = super().action_post()
+        for invoice in self:
+            if invoice.move_type == "in_invoice":
+                for line in invoice.invoice_line_ids.filtered(
+                    lambda l: not l.stock_valuation_ok
+                ):
+                    line._update_product_standard_price()
+                    line._create_in_svl()
+        return res
+
+
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    stock_valuation_ok = fields.Boolean(
+        string="Stock valuation is ok ?", compute="_compute_stock_valuation_ok"
+    )
+
+    @api.depends("price_unit", "purchase_line_id", "purchase_line_id.price_unit")
+    def _compute_stock_valuation_ok(self):
+        for line in self:
+            stock_valuation_ok = True
+            svl = (
+                self.env["stock.valuation.layer"]
+                .sudo()
+                .search(
+                    [
+                        ("product_id", "=", line.product_id.id),
+                        ("account_move_id", "=", line.move_id.id),
+                        ("company_id", "=", line.company_id.id),
+                    ]
+                )
+            )
+            if (
+                not svl
+                and line.purchase_line_id
+                and line.price_unit != line.purchase_line_id.price_unit
+            ):
+                stock_valuation_ok = False
+            line.stock_valuation_ok = stock_valuation_ok
```

Comment:
avoid add if when using a boolean

```suggestion
           line.stock_valuation_ok = (
               not svl
               and line.purchase_line_id
               and line.price_unit != line.purchase_line_id.price_unit
            )
```